attestation/slsa/slsa.go

```diff
@@ -0,0 +1,241 @@
+		switch name := attestor.Attestor.Name(); name {
+		// Pre-material Attestors
+		case environment.Name:
+			envs := attestor.Attestor.(*environment.Attestor).Variables
+			pbEnvs := make(map[string]interface{}, len(envs))
+			for name, value := range envs {
+				pbEnvs[name] = value
+			}
+
+			internalParameters["env"] = pbEnvs
+
+		case git.Name:
+			digestSet := attestor.Attestor.(git.GitAttestor).Data().CommitDigest
+			remotes := attestor.Attestor.(git.GitAttestor).Data().Remotes
+			digests, _ := digestSet.ToNameMap()
+
+			for _, remote := range remotes {
+				p.PbProvenance.BuildDefinition.ResolvedDependencies = append(
+					p.PbProvenance.BuildDefinition.ResolvedDependencies,
+					&v1.ResourceDescriptor{
+						Name:   remote,
+						Digest: digests,
+					})
+			}
+
+		case github.Name:
+			gh := attestor.Attestor.(github.GitHubAttestor)
+			p.PbProvenance.RunDetails.Builder.Id = GHABuilderId
+			p.PbProvenance.RunDetails.Metadata.InvocationId = gh.Data().PipelineUrl
+			digest := make(map[string]string)
+			digest["sha1"] = gh.Data().JWT.Claims["sha"].(string)
+
+			p.PbProvenance.BuildDefinition.ResolvedDependencies = append(
+				p.PbProvenance.BuildDefinition.ResolvedDependencies,
+				&v1.ResourceDescriptor{
+					Name:   gh.Data().ProjectUrl,
+					Digest: digest,
+				})
+
+		case gitlab.Name:
+			gl := attestor.Attestor.(*gitlab.Attestor)
+			p.PbProvenance.RunDetails.Builder.Id = GLCBuilderId
+			p.PbProvenance.RunDetails.Metadata.InvocationId = gl.PipelineUrl
+			digest := make(map[string]string)
+			digest["sha1"] = gl.JWT.Claims["sha"].(string)
+
+			p.PbProvenance.BuildDefinition.ResolvedDependencies = append(
+				p.PbProvenance.BuildDefinition.ResolvedDependencies,
+				&v1.ResourceDescriptor{
+					Name:   gl.ProjectUrl,
+					Digest: digest,
+				})
+
+		// Material Attestors
+		case material.Name:
+			mats := attestor.Attestor.(material.MaterialAttestor).Materials()
+			for name, digestSet := range mats {
+				digests, _ := digestSet.ToNameMap()
+				p.PbProvenance.BuildDefinition.ResolvedDependencies = append(
+					p.PbProvenance.BuildDefinition.ResolvedDependencies,
+					&v1.ResourceDescriptor{
+						Name:   name,
+						Digest: digests,
+					})
+			}
+
+		// CommandRun Attestors
+		case commandrun.Name:
+			var err error
+			ep := make(map[string]interface{})
+			ep["command"] = strings.Join(attestor.Attestor.(commandrun.CommandRunAttestor).Data().Cmd, " ")
+			p.PbProvenance.BuildDefinition.ExternalParameters, err = structpb.NewStruct(ep)
+			if err != nil {
+				return err
+			}
+			// We have start and finish time at the collection level, how do we access it here?
+			p.PbProvenance.RunDetails.Metadata.StartedOn = timestamppb.New(time.Now())
+			p.PbProvenance.RunDetails.Metadata.FinishedOn = timestamppb.New(time.Now())
+
+		// Product Attestors
+		case product.ProductName:
+			maps.Copy(p.products, attestor.Attestor.(product.ProductAttestor).Products())
+
+		// Post Attestors
+		case oci.Name:
```

Comment:
- I see that we have the ProductAttestor interface https://github.com/in-toto/go-witness/blob/74a58da23ce6d08323d763807748d7584bad8ade/attestation/product/product.go#L50
- we cast OCI attestor to this interface if it is found:
```
		case oci.Name:
			maps.Copy(p.products, attestor.Attestor.(product.ProductAttestor).Products())
		}
```
- However, from what I can see the Products() function isn't implemented for this attestor :thinking_face:? https://github.com/in-toto/go-witness/blob/link-attestor/attestation/oci/oci.go

Reply:
Yeah, so this actually makes sense now that I think about it. The OCI attestor doesn't actually create any files but, it does create additional subjects for the OCI references it detects. I'll need to think through what subjects we actually want to add to the SLSA provenance attestation. 🤔

Reply:
So, for now, we're just adding all of the subjects. When we make future improvements to the OCI attestor, we should revisit this.